crosshair/libimpl/builtinslib.py

```diff
@@ -1916,6 +1916,31 @@ def find(self, substr, start=None, end=None):
         start = 0 if start is None else force_to_smt_sort(start, SmtInt)
         return SmtInt(z3.IndexOf(smt_mystr, smt_substr, start))
 
+    def rfind(self, substr, start=None, end=None):
+        space = self.statespace
+        sub = force_to_smt_sort(substr, SmtStr)
+
+        start = 0 if start is None else force_to_smt_sort(start, SmtInt)
+        end = z3.Length(value) if end is None else force_to_smt_sort(end, SmtInt)
+        value = z3.SubString(self.var, start, end - 1)
+        match_index = z3.Int("match_index")
+        return_value = z3.Int("return_value")
+        space.add(
+            return_value == z3.If(match_index == -1, -1, match_index + start)
+        )
+
+        index_remaining = match_index + z3.Length(sub)
+        last_match = z3.SubString(value, match_index, z3.Length(sub))
+        remaining = z3.SubString(
+            value, index_remaining, z3.Length(value) - index_remaining
+        )
+        found_match = z3.And(
+            z3.Contains(last_match, sub), z3.Not(z3.Contains(remaining, sub))
+        )
+        no_match = z3.And(z3.Not(z3.Contains(value, sub)), match_index == -1)
+        space.add(z3.Or(no_match, found_match))
```

Comment:
A common tension in CrossHair: we can make things easier on the solver by adding branches in the path tree with `smt_fork()`. The downside is that we have to run more iterations. But especially for sting operations, Z3 gets to unsat pretty fast.

To make that idea concrete in your example, instead of this disjunction, we might say way up at the top:

```
if space.smt_fork(z3.Contains(self.var, sub)):
  ...
else:
  return -1
```
And then you can simplify some of the Z3 expressions because you can just assume that there is at least one match. CrossHair will try one branch on one iteration and the other on a future iteration.

To be clear, either approach works - this is really about tuning.

Reply:
If we want to tune, we should have some measurements IMO. Good to know there are some options, though. I think they do make the code more readable.